crates/runtime/src/lib.rs

```diff
@@ -304,6 +306,18 @@ impl_runtime_apis! {
             Starknet::nonce(address)
         }
 
+        fn events() -> Vec<EventWrapper> {
+            // (Greg) Substrate documentation states "Should only be called if you know
+            // what you are doing and outside of the runtime block
+            // execution". Is it ok to call here?
+            System::read_events_no_consensus().filter_map(|event| {
```

Comment:
So I stumbled upon this

> Please note that it's strongly recommended to NOT read the event records in the runtime. Events can grow very large (think about there were >20mb blocks on Kusama), and the read operation is very expensive. So when you read the event, you essentially read all the events from the storage at once, and iterate over them one by one. Instead, you can create your own vector storage item to track the event you are interested.

Tho, here we don't do it in the runtime per se so should be fine ? I guess we'd see it from the benchmarks anyway

Reply:
I'm still blurry on some substrate concepts, but when you query the Runtime from the runtime API, doesn't that cause some workload on your runtime and slow it down? Hence querying large amount of events is maybe not recommended?

Reply:
There's no option to read the events between block k to n ? Because indeed it'll grow very very fast. Argent accounts emit events at each call (inside the account) + fee transfer emits also an event

Reply:
This call is done in the runtime API, so I think it only reads the events for one block, since when calling a function from the runtime API, you always have to provide a `block_hash`. But it could still be very heavy.

Reply:
Also, this is the fallback method for reading the events, isn't it?
If we have the correct storage schema we will be reading directly from the storage through `overrides`

Reply:
Yes correct, this is the fallback, events are read from the `System` pallet storage.